### DISCUSSION TITLE
Elephant foot correction : erode, dimming, and... This suggestion ?

### DISCUSSION THREAD

**Discussion by Alomphega** [sn4k3/UVtools#485] [Ideas]
Hello,

I try to find a way to avoid Elephant feet in an easily but efficiently.
I can't figure out exactly what the dimming does, and the erode seems too steep to me.
Plus, AFAIK the dimming can make some issues, as yourself, Tiago, talk about in the anti aliasing wiki : 
« Remember that a faded pixel will under-cure, if you keep fading adjacent pixels the darker they get, it become less probable for them to bound to each other – Creating floating partial cured bits and milky stuff in the resin. While is not that bad isn’t that good either and this can be avoided… »

Sometimes ago, I suggested this solution below to Chitu Systems but I guess I was not clear enough, maybe. 
I paste some points of my suggestion, hoping I'll find better understanding here : 

Since I understood my issue (you can see it at https://www.facebook.com/groups/chitubox/posts/957404251586944/ )
I wonder if there could be in your agenda to improve a bit this function as this : 
Make it progressive : just following the Transition Layer Count to change the size of each layer according this transition layer count.

Nowadays the layer size change is sudden after the bottom layer count and the transition layer count as shown in my facebook discussion above in sample. 
ie: if the bottom layer count is 5 and transition layer is 10, the layer size change for the bottom tolerance compensation changes at 16th layer from bottom, which is too much sudden and make a gap on prints.
IMHO taking the same example as in facebook group discussion, just making the layer size catching up regularly from 6th to 16th layer to arrive to the normal STL file layers size could make the difference.

And plus the tolerance compensation gap avoided, the Elephant's foot issue will be definitively avoided in an easy but elegant way.

A draw to explain what I think of : 
![image](https://user-images.githubusercontent.com/27571099/170302468-52311052-bf56-461f-9cac-e691efb37208.png)

What do you think about this ?

Thank you very much for your attention reading me,
Cordially,
Guy

**Comment by sn4k3** [comment]
Does Jan method (Wait time before cure) don't solve your problem? It also include faded wait time.
What function you want to see improved? The erode method?
On erode you can already fade the erosion from x to 1, but you need to set it up, using profiles or .uvtop files you can easy set your criteria and execute it on each file. In the extreme scripting can be used.
Picking the set transition layer count is a feature which will only work on CTB and anycubic formats, it's not something everyone can benefit of

**Reply by Alomphega** [reply]
(I already had a Github account but I was too silly to think to select a category to post 😉 )
Alas Jan method does not solve my (random) lacking of bonding on plate for my plant-based translucent green resin. So I have to increase bottom exposure time, and boom, I get elephant foot.
If I can fade the erosion from x to scale 1 in accordance with the transition layers count, please tell me how I can set it, because I did not find this !
(BTW since the layers are eroded into UVTools and if UVTools allow to set the transition layer count to drive the transition of layer scaling, I do not understand why it is CTB and Anycubic formats only.)

**Reply by sn4k3** [reply]
Github is way better to keep posts and response.

>  lacking of bonding on plate

Flex plate? If not, do you sand it?

> So I have to increase bottom exposure time.

How much?

> If I can fade the erosion from x to scale 1 in accordance with the transition layers count

You can.
Go to Morph, erode, select bottom layers, and erode iterations, eg 10. Save a profile and run
Repeat last step but increase range from first normal layer to transition end layer, mark chamfer, eg: 10->1. Save a profile and run.

Now if you use always the same number of bottom layers and transition layers you can reload that profiles to save you time. 
Also you can export those as .uvtop file, drag and drop into UI to execute the operations.
As alternative a script can be made to auto detect, set and run all into one operation

> (BTW since the layers are eroded into UVTools and if UVTools allow to set the transition layer count to drive the transition of layer scaling, I do not understand why it is CTB and Anycubic formats only.)

Only CTB and derivants and Photon formats have transition layer count defined on file format. So other formats i can't know for sure how many transition layers are because formats lack of that information.
UVtools allow to set on most formats because it set directly to layers, but it will not keep that info on file itself, so next time you reload the file, the transition layer count info is lost.
Also note some formats have hardware transition layer count, CTB is software while Anycubic is hardware.
Those very specific things will need to do manual or via script, the tools will always have generic and common usage, ie, they are not suggestions or automations.

BTW, dimming tool also have chamfer option, you can opt for erode or dim, you can test which method fit you better.

If you want a script to do that auto chamfer based on transition layers, i can also do for you. But first test manual method to see if it works.

**Reply by Alomphega** [reply]
> Github is way better to keep posts and response.

Yes, agreed !
 
> > lacking of bonding on plate
> 
> Flex plate? If not, do you sand it?

No, I take off the flex plates. 
My plates are in good surfaces (Changed for Saturn S new plates), so this is not the issue.

> > So I have to increase bottom exposure time.
> 
> How much?

I tried so much times... From 30 seconds (original ones) to 40 seconds, but this was before my tries of Jan's advices.
Since I used wait time before cure, I tried till 35 seconds, it was not enough.
Now I'm printing trying a 40 seconds bottom layers exposure, but with your UVTools|Jan corrections. Crossing my fingers.


> > If I can fade the erosion from x to scale 1 in accordance with the transition layers count
> 
> You can. Go to Morph, erode, select bottom layers, and erode iterations, eg 10. Save a profile and run Repeat last step but increase range from first normal layer to transition end layer, mark chamfer, eg: 10->1. Save a profile and run.

Superb ! Thank you very much, I'll try ASAP tomorrow !


> Now if you use always the same number of bottom layers and transition layers you can reload that profiles to save you time. Also you can export those as .uvtop file, drag and drop into UI to execute the operations. As alternative a script can be made to auto detect, set and run all into one operation

Thanks a lot. I'll try this.

> > (BTW since the layers are eroded into UVTools and if UVTools allow to set the transition layer count to drive the transition of layer scaling, I do not understand why it is CTB and Anycubic formats only.)
> 
> Only CTB and derivants and Photon formats have transition layer count defined on file format. So other formats i can't know for sure how many transition layers are because formats lack of that information. UVtools allow to set on most formats because it set directly to layers, but it will not keep that info on file itself, so next time you reload the file, the transition layer count info is lost. Also note some formats have hardware transition layer count, CTB is software while Anycubic is hardware. Those very specific things will need to do manual or via script, the tools will always have generic and common usage, ie, they are not suggestions or automations.
> 
> BTW, dimming tool also have chamfer option, you can opt for erode or dim, you can test which method fit you better.
> 
> If you want a script to do that auto chamfer based on transition layers, i can also do for you. But first test manual method to see if it works.

You are my hero ! ;-)
I'll try ASAP and will come back for giving feedback.

Cordially,
Guy

**Comment by sn4k3** [comment]
> No, I take off the flex plates.
My plates are in good surfaces (Changed for Saturn S new plates), so this is not the issue.

Sand plate with a coarse grit make a huge difference, on my ibee it come presanded, it almost cut my finger, prints stick like a rock at 15s and i dont use any wait time before cure... Really i have a huge hard time to remove the print, now i use 10s but considering 5s atm because even 10s looks too much.
Smooth plates are the number 1 enemy

**Reply by Alomphega** [reply]
I agree, so mines aren't smooth but rough, the late Elegoo build plates are so.
I have sanded my original ones, and my second ones (ordered for this) are already rough.

On an other hand beware with Saturn we have another issue with the "Z=0 + 0,x mm" : it lacks some 0,x mm on the Z axis of the printed objects, and we have to set the Z=0 + these 0,x mm to get the right pieces dimensions (Elegoo support advises to do this).
I guess this is the first reason we have to set some more time for bottom layers exposure, to let the resin grow enough to fill the added gap.
So the need of fine-tuning for the "Elephant foot compensation"... Which I will try with UVTools today !

I know, I know, the best should be to mechanically solve this issue... But many of us do not have nor the time, nor the knowledge - nor both - for this !

See you soon,
Guy

**Reply by sn4k3** [reply]
Cant you home/calibrate Z with less or more offset to counter that problem on Z axis?

**Reply by Alomphega** [reply]
> Cant you home/calibrate Z with less or more offset to counter that problem on Z axis?

To have a Z axis dimension accurate, I have to calibrate/set this Z=0 + 0,x mm (depending of which Saturn "0,x mm" could be from 0,2 mm till 0,8 mm !)

**Reply by Alomphega** [reply]
To be more precise : the lack on Z axis within my Saturn printers is a lack of dimension on the base of the piece (build plate side), not a scale one. So I guess Z calibration ("XYZ Accuracy" Calibration menu tool) cannot help since it work on scale, is'nt it ?

About my "Morphantfoot" tries : 
I guess with the help of the calculator with these values : 
![image](https://user-images.githubusercontent.com/27571099/170499320-2fcbba1c-c40f-4c4c-97e6-e10d8b252644.png)
that with 10 pixels iterations for the Morph+Erode tool like in your example I cut off 0,5mm from the boundaries of the layers, so for instance with this, I cut off 1mm on a circle diameter. 
Am I right in my understanding of your tool ?

**Reply by sn4k3** [reply]
> 10 pixels iterations .. I cut off 0,5mm from the boundaries of the layers, so for instance with this, I cut off 1mm on a circle diameter.

Correct.
With large iteration numbers i recommend you use "Dynamic Kernel" check under "Kernel" expander.
It will make better contours

**Reply by Alomphega** [reply]
> With large iteration numbers i recommend you use "Dynamic Kernel" check under "Kernel" expander.
> It will make better contours

I saw it, but I don't figure out what it is nor how using it ! 🤔

**Comment by Alomphega** [comment]
Hello,

Quite a success ! 
We need to fine-tune a bit, but the result is very encouraging as the process is very simple.

Here are 2 (poor) photos to let you see the difference between before with Chitubox Bottom Tolerance Compensation, and after with my "UVTools Morphantfeet" first try (without fine-tuning) : 

Before (chamfer is not chamfer at all, just waves)
![image](https://user-images.githubusercontent.com/27571099/170654432-fb903362-232b-4e9d-a74f-80cab444da84.png)

After (chamfer looks much more like a chamfer)
![image](https://user-images.githubusercontent.com/27571099/170654519-76c1cb39-8cbc-4acb-8e7c-09ba7ef7251f.png)

Today is the time to fine tune, I will come back to give a bit more feedback.

Cordially,
Guy

**Reply by sn4k3** [reply]
Big difference... UVtools is such a powerfull set ;)

**Reply by Alomphega** [reply]
👍Yes it is amazing ! I wonder who is the brilliant man who developed it ! ;)

More seriously, now I'm printing trying even without the bottom layers erode step, just the chamfer one from first layer to the last transition one, and a bit less correction (0,4mm instead of 0,5mm). Wait and see !

**Reply by sn4k3** [reply]
> 👍Yes it is amazing ! I wonder who is the brilliant man who developed it ! ;)

Who knows... Tales are the man is not from this planet 👽

> More seriously, now I'm printing trying even without the bottom layers erode step, just the chamfer one from first layer to the last transition one, and a bit less correction (0,4mm instead of 0,5mm). Wait and see !

It will take some trial-error. No pain no gain...

**Reply by Alomphega** [reply]
> It will take some trial-error. No pain no gain...

Hurrah, since all the pains I already own, probably I gained so much that now I'm so a rich man than I do not even knew this (because we don't care about what we don't need ;) )
